New idiom: return moved arg on error (#336)
rust-unofficial/patterns: idioms/return-moved-arg-on-error.md

# A succinct name for the pattern

## Description

> If fallible method receives consumes argument, return it back inside error.

Reviewer: You need to drop either `receives` or `consumes`. And maybe add a few `a`s and `an`s. Also, we're usually talking about `function`s rather than `methods`. Suggestion:

```suggestion
If a fallible function consumes an argument, return that argument back inside an error.
```

Another thing which you may want to highlight is that you usually do this for functions mutating or converting some operand.

## Example

```rust
pub fn send(value: String) -> Result<(), SendError> {
    // Simulate non-deterministic fallible action.
    use std::time::SystemTime;
    let period = SystemTime::now().duration_since(SystemTime::UNIX_EPOCH).unwrap();
    if period.subsec_nanos() % 2 == 1 {
        Ok(())
    } else {
        Err(SendError(value))
    }
}

pub struct SendError(String);

fn main() {
    use std::mem::take;

    let mut value = "imagine this is very long string".to_string();

    let success = 'send: {
        // Try to send value two times.
        for _ in 0..2 {
            value = match send(take(&mut value)) {
                Ok(()) => break 'send true,
                Err(SendError(value)) => value,
            }
        }
        false
    };
    
    println!("success: {}", success);
}
```

## Motivation

> In case of error you may want to try some alternative way or to

Reviewer: ```suggestion
In case of an error you may want to try some alternative way or to
```

> retry action in case of non-deterministic function. But if argument

Reviewer: ```suggestion
retry action in case of non-deterministic function. But if the argument
```

is always consumed, you are forced to clone it on every call, which
is not very efficient.

> Standard library uses this approach in e.g. `String::from_utf8` method.

Reviewer: ```suggestion
The standard library uses this approach in e.g. `String::from_utf8` method.
```

> When given vector doesn't contain valid UTF-8, `FromUtf8Error` is returned.

Reviewer: ```suggestion
When given a vector that doesn't contain valid UTF-8, a `FromUtf8Error` is returned.
```

# A succinct name for the pattern

## Description

If a fallible function consumes an argument, return that argument back inside
an error.

## Example

```rust
pub fn send(value: String) -> Result<(), SendError> {

>     // Simulate non-deterministic fallible action.

Reviewer 2: ```suggestion
    println!("using {value} in a meaningful way");
    // Simulate non-deterministic fallible action.
```
We could write something like this to make it explicit that the function is using the value. Actually, it would be great if we could think of an example where the value is used for real.

    use std::time::SystemTime;
    let period = SystemTime::now().duration_since(SystemTime::UNIX_EPOCH).unwrap();
    if period.subsec_nanos() % 2 == 1 {
        Ok(())
    } else {
        Err(SendError(value))
    }
}

pub struct SendError(String);

fn main() {
    use std::mem::take;

    let mut value = "imagine this is very long string".to_string();

    let success = 'send: {
        // Try to send value two times.
        for _ in 0..2 {

>             value = match send(take(&mut value)) {

Reviewer 2: why do you use `take` here?
The code works even without it for me. 🤔
